File: 프로그래머스/3/87694. 아이템 줍기/아이템 줍기.py
from collections import deque

def solution(rectangle, characterX, characterY, itemX, itemY):
    arr = [[-1 for _ in range(102)] for _ in range(102)]
    visited = [[1 for _ in range(102)] for _ in range(102)]

    for r in rectangle:
        x1, y1, x2, y2 = map(lambda x: x * 2, r)

        for i in range(x1, x2 + 1):
            for j in range(y1, y2 + 1):
                if x1 < i < x2 and y1 < j < y2:
                    arr[i][j] = 0
                elif arr[i][j] != 0:
                    arr[i][j] = 1

    sx, sy, ex, ey = 2 * characterX, 2 * characterY, 2 * itemX, 2 * itemY
    que = deque([(sx, sy)])
    answer = 0
    d = [(0, 1), (1, 0), (0, -1), (-1, 0)]

    while que:
        x, y = que.popleft()

        if x == ex and y == ey:
            answer = visited[x][y] // 2
            break

        for k in range(4):
            nx, ny = x + d[k][0], y + d[k][1]

            if arr[nx][ny] == 1 and visited[nx][ny] == 1:
                visited[nx][ny] += visited[x][y]
                que.append((nx, ny))

    return answer

File: 프로그래머스/3/87694. 아이템 줍기/아이템 줍기.java
import java.util.*;

class Solution {
    public int solution(int[][] rectangle, int characterX, int characterY, int itemX, int itemY) {
        int[][] arr = new int[102][102];

        for (int[] r : rectangle) {
            int x1 = r[0] * 2;
            int y1 = r[1] * 2;
            int x2 = r[2] * 2;
            int y2 = r[3] * 2;

            for (int i = x1; i <= x2; i++) {
                for (int j = y1; j <= y2; j++) {
                    if (x1 < i && i < x2 && y1 < j && j < y2) {
                        arr[i][j] = 1;
                    } else if (arr[i][j] != 1) {
                        arr[i][j] = 2;
                    }
                }
            }
        }

        Queue<int[]> que = new LinkedList<>();
        int[][] visited = new int[102][102];
        int sx = characterX * 2;
        int sy = characterY * 2;
        int ex = itemX * 2;
        int ey = itemY * 2;
        int[][] dirs = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};

        que.offer(new int[]{sx, sy});
        
        for (int i = 0; i < 102; i++) {
            Arrays.fill(visited[i], 1);
        }

        while (true) {
            int[] temp = que.poll();
            int x = temp[0];
            int y = temp[1];

            if (x == ex && y == ey) {
                return visited[x][y] / 2;
            }

            for (int[] dir : dirs) {
                int nx = x + dir[0];
                int ny = y + dir[1];

                if (arr[nx][ny] == 2 && visited[nx][ny] == 1) {
                    que.offer(new int[]{nx, ny});
                    visited[nx][ny] += visited[x][y];
                }
            }
        }
    }
}